File: 10week/LeetCode-17-Letter-Combinations-of-a-Phone-Number.py
class Solution:    
    def letterCombinations(self, digits: str) -> list[str]:
        def dfs(index, path):
            if len(path) == len(digits):
                result.append(path)
                return
            
            for i in range(index, len(digits)):
                for j in dict[digits[i]]:
                    dfs(i + 1, path + j)
   
        dict = {
            "2": "abc", "3": "def", "4": "ghi", "5": "jkl",
            "6": "mno", "7": "pqrs", "8": "tuv", "9": "wxyz"
        }
        if not digits:
            return []
        
        result = []
        dfs(0, "")

        return result

File: graph/Leetcode-17-LetterCombinationsOfAPhoneNumber/Leetcode-17-LetterCombinationsOfAPhoneNumber.cpp
class Solution {
    map<char, string> phone = {make_pair('2', "abc"), make_pair('3', "def"), make_pair('4', "ghi"), make_pair('5', "jkl"), make_pair('6', "mno"), make_pair('7', "pqrs"), make_pair('8', "tuv"), make_pair('9', "wxyz")};

public:
    vector<string> letterCombinations(string digits) {
        vector<string> result;
        myDfs(digits, 0, "", result);

        return result;
    }

    void myDfs(string digits, int index, string myStr, vector<string>& result){
        string target = phone[digits[index]];
        for (char toAppend : target) {
            string newStr(myStr);
            newStr.push_back(toAppend);
            if( index == digits.size()-1 ) result.push_back(newStr);
            else myDfs(digits, index+1, newStr, result);
        }
    }
};